include/misc/speculation.h

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2019 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_MISC_SPECULATION_H
+#define ZEPHYR_MISC_SPECULATION_H
+
+#include <zephyr/types.h>
+
+/**
+ * Sanitize an array index against bounds check bypass attacks aka the
+ * Spectre V1 vulnerability.
+ *
+ * CPUs with speculative execution may speculate past any size checks and
+ * leak confidential data due to analyzsis of micro-architectural properties.
+ * This will unconditionally truncate any out-of-bounds indexes to
+ * zero in the speculative execution path using bit twiddling instead of
+ * any branch instructions.
+ *
+ * Example usage:
+ *
+ * if (index < size) {
+ *     index = k_array_index_safe(index, size);
+ *     data = array[index];
+ * }
+ *
+ * @param index Untrusted array index which has been validated, but not used
+ * @param size Size of the array
+ * @return The original index value if < size, or 0
+ */
+static inline u32_t k_array_index_safe(u32_t index, u32_t array_size)
```

Comment:
I don't like using the word "safe" in APIs, because their meaning wears off quickly, giving a false sensation of safety -- especially since that can mean a lot of different things in different contexts.

Might be better to just use "nospec" like Linux does here.

Reply:
OK